File: app/src/main/java/kr/com/misemung/MiseMungGApplication.java
package kr.com.misemung;

import android.app.Activity;
import android.content.Context;
import android.content.res.Configuration;
import android.os.Build;
import android.os.Bundle;
import android.os.Environment;
import android.support.multidex.MultiDex;
import android.support.multidex.MultiDexApplication;
import android.text.TextUtils;
import android.util.Log;

import com.google.firebase.analytics.FirebaseAnalytics;

import java.io.File;
import java.util.ArrayList;
import java.util.Collections;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

import io.realm.Realm;
import io.realm.RealmConfiguration;
import kr.com.misemung.realm.entity.AirRecord;


/**
 * 미세멍지 application
 */
public class MiseMungGApplication extends MultiDexApplication {
	private static final String TAG = MiseMungGApplication.class.getSimpleName();
	private static MiseMungGApplication ourApplication;

	private static Context mContext;

	public Activity nowActivity;	//현재 실행되는 액티비티(viewer 여부)

    private FirebaseAnalytics mAnalytics;

    public static MiseMungGApplication Instance()
	{
		return ourApplication;
	}
	public MiseMungGApplication()
	{
		ourApplication = this;
	}

	public static Activity currentActivity = null;

	public void onCreate() {
		super.onCreate();
		Log.i(TAG, "MiseMungGApplication Create");

		mContext = getApplicationContext();

		Realm.init(mContext);

		try {
			RealmConfiguration newConfig = new RealmConfiguration.Builder()
					.name("misemung.realm")
					.schemaVersion(1)
					.deleteRealmIfMigrationNeeded()
					.compactOnLaunch()
					.build();
			Realm.setDefaultConfiguration(newConfig);

			Realm realm = Realm.getDefaultInstance();
			RealmConfiguration config = realm.getConfiguration();

			Log.e("MiseMungGApplication", "REALM path:" + config.getPath());
			File realmFile = new File(config.getPath());
			if (!realmFile.exists()) {
				realm.createObject(AirRecord.class);
			}

		} catch (RuntimeException ignore) {}

		//Thread.setDefaultUncaughtExceptionHandler(new EBookUncaughtExceptionHandler(getBaseContext(), getContentResolver(), Thread.getDefaultUncaughtExceptionHandler()));

	}

	protected void attachBaseContext(Context base) {
		super.attachBaseContext(base);
		MultiDex.install(this);
	}

	@Override
	public void onConfigurationChanged(Configuration newConfig) {
		super.onConfigurationChanged(newConfig);
		Log.d(TAG, "Application onConfigurationChanged fontScale: " + newConfig.fontScale);
		Log.d(TAG, "Application onConfigurationChanged keyboardHidden: " + newConfig.keyboardHidden);
		Log.d(TAG, "Application onConfigurationChanged locale: "+newConfig.locale);
		Log.d(TAG, "Application onConfigurationChanged orientation: " + newConfig.orientation);
		Log.d(TAG, "Application onConfigurationChanged screenLayout: " + newConfig.screenLayout);
	}

	public void onLowMemory() {
		super.onLowMemory();
		Log.d(TAG, "Application onLowMemory");
	}

	public void setNowActivity(Activity activity) {
		MiseMungGApplication.Instance().nowActivity = activity;
	}

	synchronized  public FirebaseAnalytics getDefaultAnalytics() {
		if(mAnalytics == null) {
			mAnalytics = FirebaseAnalytics.getInstance(this);
		}
		return mAnalytics;
	}

	public void setLogFirebaseAnalytics(String gubun) {
		if (mAnalytics == null) {
			getDefaultAnalytics();
		}

		mAnalytics.logEvent(gubun, new Bundle());
	}

	public Activity getCurrentActivity() {
		return MiseMungGApplication.currentActivity;
	}

	public static void setCurrentActivity(Activity currentActivity) {
		MiseMungGApplication.currentActivity = currentActivity;
	}

	/** Dir Database */
	public String getAppDatabaseDir(){
		return getDatabasePath("sbooks.db").getAbsolutePath();
	}
	/** Dir File */
	public String getAppFileDir(){
		return getFilesDir().getAbsolutePath();
	}
	/** Dir Cache */
	public String getAppCacheDir(){
		if(getExternalCacheDir() == null)
			return getCacheDir().getAbsolutePath();
		else
			return getExternalCacheDir().getAbsolutePath();
	}
	/** Dir Cert */
	public String getAppDRMCertDir(){
		return new File( getAppDataDir(), "cert").getAbsolutePath();
	}
	/** Dir Lic */
	public String getAppDRMLicDir(){
		return new File( getAppDataDir(), "lic").getAbsolutePath();
	}
	/** Dir SharedPreferences */
	public String getAppSharedPreferencesDir(){
		return new File( getAppDataDir(), "shared_prefs").getAbsolutePath();
	}
	/** Dir Data */
	public String getAppDataDir(){
		return getFilesDir().getParent();
	}

	public static String[] getStorageDirectories() {
		String[] storageDirectories;
		String rawSecondaryStoragesStr = System.getenv("SECONDARY_STORAGE");

		if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.KITKAT) {
			List<String> results = new ArrayList<String>();
			File[] externalDirs = mContext.getExternalFilesDirs(null);
			if (externalDirs != null) {
				for (File file : externalDirs) {
					if (file == null) {
						continue;
					}
					String path = file.getPath().split("/Android")[0];
					if((Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP && Environment.isExternalStorageRemovable(file))
							|| rawSecondaryStoragesStr != null && rawSecondaryStoragesStr.contains(path)){
						results.add(path);
					}
				}
				storageDirectories = results.toArray(new String[0]);
			} else {
				storageDirectories = null;
			}
		} else {
			final Set<String> rv = new HashSet<String>();

			if (!TextUtils.isEmpty(rawSecondaryStoragesStr)) {
				final String[] rawSecondaryStorages = rawSecondaryStoragesStr.split(File.pathSeparator);
				Collections.addAll(rv, rawSecondaryStorages);
			}
			storageDirectories = rv.toArray(new String[rv.size()]);
		}
		return storageDirectories;
	}
}

File: app/src/main/java/kr/com/misemung/MiseMungGApplication.kt
package kr.com.misemung

import android.app.Activity
import android.content.Context
import android.content.res.Configuration
import android.os.Build
import android.os.Bundle
import android.os.Environment
import android.text.TextUtils
import android.util.Log
import androidx.multidex.MultiDex
import androidx.multidex.MultiDexApplication
import com.google.firebase.analytics.FirebaseAnalytics
import io.realm.Realm
import io.realm.RealmConfiguration
import kr.com.misemung.realm.entity.AirRecord
import java.io.File
import java.util.*

/**
 * 미세멍지 application
 */
class MiseMungGApplication : MultiDexApplication() {
    var nowActivity //현재 실행되는 액티비티(viewer 여부)
            : Activity? = null
    private var mAnalytics: FirebaseAnalytics? = null
    override fun onCreate() {
        super.onCreate()
        Log.i(TAG, "MiseMungGApplication Create")
        mContext = applicationContext
        Realm.init(mContext)
        try {
            val newConfig = RealmConfiguration.Builder()
                .name("misemung.realm")
                .schemaVersion(1)
                .allowWritesOnUiThread(true)
                .deleteRealmIfMigrationNeeded()
                .compactOnLaunch()
                .build()
            Realm.setDefaultConfiguration(newConfig)
            val realm = Realm.getDefaultInstance()
            val config = realm.configuration
            Log.e("MiseMungGApplication", "REALM path:" + config.path)
            val realmFile = File(config.path)
            if (!realmFile.exists()) {
                realm.createObject(AirRecord::class.java)
            }
        } catch (ignore: RuntimeException) {
            Log.e("MiseMungGApplication", ignore.toString())
        }

        //Thread.setDefaultUncaughtExceptionHandler(new EBookUncaughtExceptionHandler(getBaseContext(), getContentResolver(), Thread.getDefaultUncaughtExceptionHandler()));
    }

    override fun attachBaseContext(base: Context) {
        super.attachBaseContext(base)
        MultiDex.install(this)
    }

    override fun onConfigurationChanged(newConfig: Configuration) {
        super.onConfigurationChanged(newConfig)
        Log.d(TAG, "Application onConfigurationChanged fontScale: " + newConfig.fontScale)
        Log.d(TAG, "Application onConfigurationChanged keyboardHidden: " + newConfig.keyboardHidden)
        Log.d(TAG, "Application onConfigurationChanged locale: " + newConfig.locale)
        Log.d(TAG, "Application onConfigurationChanged orientation: " + newConfig.orientation)
        Log.d(TAG, "Application onConfigurationChanged screenLayout: " + newConfig.screenLayout)
    }

    override fun onLowMemory() {
        super.onLowMemory()
        Log.d(TAG, "Application onLowMemory")
    }

    @JvmName("setNowActivity1")
    fun setNowActivity(activity: Activity?) {
        Instance().nowActivity = activity
    }

    @get:Synchronized
    val defaultAnalytics: FirebaseAnalytics
        get() {
            if (mAnalytics == null) {
                mAnalytics = FirebaseAnalytics.getInstance(this)
            }
            return mAnalytics!!
        }

    fun setLogFirebaseAnalytics(gubun: String?) {
        if (mAnalytics == null) {
            defaultAnalytics
        }
        mAnalytics!!.logEvent(gubun!!, Bundle())
    }

    val currentActivity: Activity?
        get() = Companion.currentActivity

    /** Dir Database  */
    val appDatabaseDir: String
        get() = getDatabasePath("sbooks.db").absolutePath

    /** Dir File  */
    val appFileDir: String
        get() = filesDir.absolutePath

    /** Dir Cache  */
    val appCacheDir: String
        get() = if (externalCacheDir == null) cacheDir.absolutePath else externalCacheDir!!.absolutePath

    /** Dir Cert  */
    val appDRMCertDir: String
        get() = File(appDataDir, "cert").absolutePath

    /** Dir Lic  */
    val appDRMLicDir: String
        get() = File(appDataDir, "lic").absolutePath

    /** Dir SharedPreferences  */
    val appSharedPreferencesDir: String
        get() = File(appDataDir, "shared_prefs").absolutePath

    /** Dir Data  */
    val appDataDir: String
        get() = filesDir.parent

    companion object {
        private val TAG = MiseMungGApplication::class.java.simpleName
        private lateinit var ourApplication: MiseMungGApplication
        private var mContext: Context? = null
        fun Instance(): MiseMungGApplication {
            return ourApplication
        }

        var currentActivity: Activity? = null
        @JvmName("setCurrentActivity1")
        fun setCurrentActivity(currentActivity: Activity?) {
            Companion.currentActivity = currentActivity
        }

        val storageDirectories: Array<String>?
            get() {
                val storageDirectories: Array<String>?
                val rawSecondaryStoragesStr = System.getenv("SECONDARY_STORAGE")
                storageDirectories = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.KITKAT) {
                    val results: MutableList<String> = ArrayList()
                    val externalDirs = mContext!!.getExternalFilesDirs(null)
                    if (externalDirs != null) {
                        for (file in externalDirs) {
                            if (file == null) {
                                continue
                            }
                            val path = file.path.split("/Android").toTypedArray()[0]
                            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP && Environment.isExternalStorageRemovable(
                                    file
                                )
                                || rawSecondaryStoragesStr != null && rawSecondaryStoragesStr.contains(
                                    path
                                )
                            ) {
                                results.add(path)
                            }
                        }
                        results.toTypedArray()
                    } else {
                        null
                    }
                } else {
                    val rv: Set<String> = HashSet()
                    if (!TextUtils.isEmpty(rawSecondaryStoragesStr)) {
                        val rawSecondaryStorages =
                            rawSecondaryStoragesStr!!.split(File.pathSeparator).toTypedArray()
                        //Collections.addAll(rv, *rawSecondaryStorages)
                    }
                    rv.toTypedArray()
                }
                return storageDirectories
            }
    }

    init {
        ourApplication = this
    }
}